### ISSUE TITLE
[Feature] Add Early Stopping

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2667]
## Problem

Training often continues after validation metric stops improving, wasting time and potentially overfitting.

### 1. Early Stopping Class

```mojo
# shared/training/early_stopping.mojo
struct EarlyStopping:
    """Stop training when monitored metric stops improving."""
    var patience: Int
    var min_delta: Float32
    var mode: String  # "min" or "max"
    var best_value: Float32
    var counter: Int
    var best_epoch: Int
    
    fn __init__(out self, patience: Int = 10, min_delta: Float32 = 0.0, mode: String = "min"):
        """
        Args:
            patience: Number of epochs to wait for improvement
            min_delta: Minimum change to qualify as improvement
            mode: "min" for loss (lower is better), "max" for accuracy (higher is better)
        """
        self.patience = patience
        self.min_delta = min_delta
        self.mode = mode
        self.best_value = Float32.infinity() if mode == "min" else -Float32.infinity()
        self.counter = 0
        self.best_epoch = 0
    
    fn step(mut self, current_value: Float32, epoch: Int) -> Bool:
        """Check if should stop training.
        
        Returns:
            True if should stop, False otherwise
        """
        var improved = False
        
        if self.mode == "min":
            improved = current_value < (self.best_value - self.min_delta)
        else:
            improved = current_value > (self.best_value + self.min_delta)
        
        if improved:
            self.best_value = current_value
            self.best_epoch = epoch
            self.counter = 0
            return False  # Don't stop
        else:
            self.counter += 1
            
            if self.counter >= self.patience:
                print("Early stopping triggered!")
                print("Best value:", self.best_value, "at epoch", self.best_epoch)
                return True  # Stop training
            else:
                print("No improvement for", self.counter, "epochs (patience:", self.patience, ")")
                return False
    
    fn get_best_epoch(self) -> Int:
        """Get epoch with best metric."""
        return self.best_epoch
```

### 2. Usage in Training Loop

```mojo
fn train():
    var model = LeNet5()
    var optimizer = Adam(model.parameters(), lr=0.001)
    var early_stopping = EarlyStopping(patience=10, min_delta=0.001, mode="min")
    var checkpoint_mgr = CheckpointManager("checkpoints/lenet5")
    
    for epoch in range(max_epochs):
        # Training
        var train_loss = train_epoch(model, optimizer)
        
        # Validation
        var val_loss = validate(model)
        
        print("Epoch", epoch, "| Train Loss:", train_loss, "| Val Loss:", val_loss)
        
        # Save checkpoint
        checkpoint_mgr.save(model, optimizer, epoch, {"val_loss": val_loss})
        
        # Check early stopping
        if early_stopping.step(val_loss, epoch):
            print("Stopping early at epoch", epoch)
            break
    
    # Restore best model
    var best_epoch = early_stopping.get_best_epoch()
    print("Restoring model from epoch", best_epoch)
    checkpoint_mgr.load_specific_epoch(model, best_epoch)
```

### 3. With Multiple Metrics

```mojo
struct EarlyStoppingMultiMetric:
    """Early stopping based on multiple metrics."""
    var stoppers: Dict[String, EarlyStopping]
    
    fn __init__(out self):
        self.stoppers = Dict[String, EarlyStopping]()
    
    fn add_metric(mut self, name: String, patience: Int, mode: String):
        """Add a metric to monitor."""
        self.stoppers[name] = EarlyStopping(patience, mode=mode)
    
    fn step(mut self, metrics: Dict[String, Float32], epoch: Int) -> Bool:
        """Check all metrics. Stop if ANY triggers."""
        for name, stopper in self.stoppers.items():
            if name in metrics:
                if stopper.step(metrics[name], epoch):
                    print("Early stopping triggered by metric:", name)
                    return True
        return False

# Usage
var early_stopping = EarlyStoppingMultiMetric()
early_stopping.add_metric("val_loss", patience=10, mode="min")
early_stopping.add_metric("val_accuracy", patience=15, mode="max")

for epoch in range(max_epochs):
    var metrics = {
        "val_loss": validate_loss(model),
        "val_accuracy": validate_accuracy(model)
    }
    
    if early_stopping.step(metrics, epoch):
        break
```

### 4. With Model Restoration

```mojo
struct EarlyStoppingWithRestore(EarlyStopping):
    """Early stopping that automatically saves/restores best model."""
    var best_model_state: Optional[Dict[String, ExTensor]]
    
    fn step(mut self, model: Model, current_value: Float32, epoch: Int) -> Bool:
        """Step and save model if improved."""
        var should_stop = super().step(current_value, epoch)
        
        # Save best model state
        if self.counter == 0:  # Just improved
            self.best_model_state = model.state_dict()
            print("Saved best model at epoch", epoch)
        
        # Restore best model if stopping
        if should_stop and self.best_model_state:
            print("Restoring best model from epoch", self.best_epoch)
            model.load_state_dict(self.best_model_state.value())
        
        return should_stop

# Usage
var early_stopping = EarlyStoppingWithRestore(patience=10, mode="min")

for epoch in range(max_epochs):
    train_epoch(model, optimizer)
    var val_loss = validate(model)
    
    # Automatically saves/restores best model
    if early_stopping.step(model, val_loss, epoch):
        break
```

### 5. Configurable from CLI

```mojo
fn main():
    var args = parse_args()
    
    var early_stopping = EarlyStopping(
        patience=args.get_int("patience", default=10),
        min_delta=args.get_float("min_delta", default=0.0),
        mode=args.get_string("mode", default="min")
    )
    
    # Train with early stopping
    train(model, optimizer, early_stopping)

# Usage:
# mojo run train.mojo --patience 15 --min_delta 0.001 --mode min
```

### 6. Logging and Analysis

```mojo
fn train_with_logging():
    var early_stopping = EarlyStopping(patience=10, mode="min")
    var history = List[Float32]()
    
    for epoch in range(max_epochs):
        var val_loss = validate(model)
        history.append(val_loss)
        
        if early_stopping.step(val_loss, epoch):
            print("\nTraining Summary:")
            print("- Total epochs:", epoch)
            print("- Best epoch:", early_stopping.get_best_epoch())
            print("- Best value:", early_stopping.best_value)
            print("- Stopped due to no improvement for", early_stopping.patience, "epochs")
            
            # Plot training history
            plot_training_curve(history, early_stopping.get_best_epoch())
            break
```

## Benefits

- Save time by stopping when no longer improving
- Prevent overfitting (stop before validation gets worse)
- Automatic hyperparameter tuning (find optimal epoch count)
- Standard ML practice

## Typical Settings

| Scenario | Patience | Min Delta | Mode |
|----------|----------|-----------|------|
| **Small dataset** | 5-10 | 0.001 | min (loss) |
| **Large dataset** | 10-20 | 0.0001 | min (loss) |
| **Classification** | 10-15 | 0.01 | max (accuracy) |
| **Regression** | 10-20 | 0.001 | min (MSE) |

## Impact

- **Severity**: LOW
- **Efficiency**: Save training time
- **Quality**: Prevent overfitting

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P1

### Depends On
- #2664
- #2665

### Blocks
- #2728
- #2742
- #2743

### Analysis
Early stopping depends on checkpointing (#2664) to save best model and learning rate scheduling (#2665) for complete training pipeline. Blocks training-related work.

---
*Analyzed by Claude Code dependency analyzer*
